enhancements/update/accepted-risks.md

```diff
@@ -0,0 +1,521 @@
+- adds a new field 'clusterversion.status.conditionalUpdates.riskNames': It
+  contains the names of risks for the conditional update. It deprecates
+  `clusterversion.status.conditionalUpdates.risks`.
+- adds a new field 'clusterversion.status.conditionalUpdateRisks.conditions': It
+  contains the observations of the conditional update risk's current status.
+
+For example,
+
+```console
+$ oc clusterversion version -o yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  channel: candidate-4.18
+  clusterID: 1c182977-5663-428d-92a3-3d2bdf3fffb6
+  desiredUpdate:
+    accepted:
+    - DualStackNeedsController
+    - LeakedMachineConfigBlocksMCO
+    - OldBootImagesPodmanMissingAuthFlag
+    force: false
+    version: 4.18.15
+status:
+  conditionalUpdateRisks:  # include every risk in the conditional updates (moved up and renamed)
+  - name: DualStackNeedsController
+    message: Upgrade can get stuck on clusters that use multiple networks together with dual stack.
+    url: https://issues.redhat.com/browse/SDN-3996
+    matchingRules:
+    - type: Always
+    conditions:
+    - status: True  # Apply=True if the risk is applied to the current cluster
+      type: Applies
+      reason: MatchingRule
+      message: The matchingRules[0] matches
+      lastTransitionTime: 2021-09-13T17:03:05Z
+  - name: LeakedMachineConfigBlocksMCO
+    message: MCO update might be blocked if some condition holds
+    url: https://issues.redhat.com/browse/SDN-2212
+    matchingRules:
+    - type: PromQL
+      promql:
+        promql: prom_query
+    conditions:
+    - status: False
+      type: Applies
+      reason: NoMatchingRule
+      message: None of the matching rules matches
+      lastTransitionTime: 2021-09-13T17:03:05Z
+  - name: DualStackNeedsController
+    ...
+  - name: RHELKernelHighLoadIOWait
+    ...
+  conditionalUpdates:
+  - release:
+      version: 4.18.16
+      image: quay.io/openshift-release-dev/ocp-release@sha256:abc123
+    riskNames:
+    - DualStackNeedsController
+    - OldBootImagesPodmanMissingAuthFlag
+    - RHELKernelHighLoadIOWait
+    risks:  # deprecated by riskNames
+  - release:
+      version: 4.18.17
+    riskNames:
+    - OldBootImagesPodmanMissingAuthFlag
+    risks:  # deprecated by riskNames
+  - release:
+      version: 4.19.1
+    riskNames:
+    - LeakedMachineConfigBlocksMCO
+    risks:  # deprecated by riskNames
+```
+
+When a conditional update is accepted, the names of its associated risks are going to be merged into `clusterversion.status.history.acceptedRisks` which is an existing field before this enhancement. For example, CVO's acceptance of `4.18.17` leads to `OldBootImagesPodmanMissingAuthFlag` being a part of value of `clusterversion.status.history.acceptedRisks`: The wording might look like "The target release ... is exposed to the risks [OldBootImagesPodmanMissingAuthFlag] and accepted by CVO because all of them are considered acceptable".
```

Comment:
I wonder if we want to introduce a new, structured field (list of accepted risks -- or items, see below)  instead of trying to pack everything into a script.

I am thinking whether this all leads to some abstract notion of "things that an admin accepts when updating". We now have accepted risks, --alllow-explicit-upgrade, --allow-with-warnings, --force... All these are a kind of acceptance, maybe it warrants a structured list of items that were accepted?

Reply:
Let me think about this and get back to it.

Reply:
Just to make sure if I understand your suggestion correctly.
You want a `struct` instead of the current `string` for `status.history.acceptedRisks`, right?

```console
$ oc explain clusterversion.status.history.acceptedRisks
GROUP:      config.openshift.io
KIND:       ClusterVersion
VERSION:    v1

FIELD: acceptedRisks <string>


DESCRIPTION:
    acceptedRisks records risks which were accepted to initiate the update.
    For example, it may menition an Upgradeable=False or missing signature
    that was overriden via desiredUpdate.force, or an update that was
    initiated despite not being in the availableUpdates set of recommended
    update targets.
```

Reply:
Good idea. I will treat this one as feature improvement and not a blocker for this enhancement.